### PR TITLE
Update local sizes for permute

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Thanks @gpolovets1 for the catch!

Update local sizes to avoid assigning the rest padding tokens to last expert using `jnp.bincount`. This accelerate 2x original perf for EP sharding, document in this [bug](https://b.corp.google.com/issues/403069924#comment38).

# Tests

Expect this test to pass: https://github.com/AI-Hypercomputer/maxtext/blob/315e551e5942b24656a4250dcfca986fb4135b72/MaxText/tests/moe_test.py#L423

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Review by gobbleturk — APPROVED** [review]
pretty hard to follow logic here without comments - consider including some shape information in the comments

**Review by gagika — APPROVED** [review]

**Comment by RissyRan** [comment]
> pretty hard to follow logic here without comments - consider including some shape information in the comments

Thanks for the feedback! Added some size info related to group sizes.
